test/Model/UserSpec.hs

```diff
@@ -14,61 +14,101 @@ main = hspec spec
+                        , credsIdent = "1"
+                        , credsExtra =
+                            [ ("name", "foo")
+                            , ("email", "bar@gmail.com")
+                            ]
+                        }
 
-        it "authenticates known users even if profile data is missing" $ do
-            Entity userId user <- runDB $ createUser "1"
-            let creds = Creds
-                    { credsPlugin = userPlugin user
-                    , credsIdent = userIdent user
-                    , credsExtra = []
-                    }
+                Right userId <- tryAuthentication creds
 
-            (runDB $ authenticateUser creds) `shouldReturn` Just userId
+                Just user <- runDB $ DB.get userId
+                userName user `shouldBe` "foo"
+                userEmail user `shouldBe` "bar@gmail.com"
 
-        it "does not authenticate unknown users if profile data is missing" $ do
-            let creds = Creds
-                    { credsPlugin = "github"
-                    , credsIdent = "1"
-                    , credsExtra = []
-                    }
+            it "updates an existing user's profile data" $ do
+                let creds = Creds
+                        { credsPlugin = "github"
+                        , credsIdent = "1"
+                        , credsExtra =
+                            [ ("name", "foo")
+                            , ("email", "bar@gmail.com")
+                            ]
+                        }
+                userId <- runDB $ do
+                    Entity planId _ <- createFreePlan
+
+                    insert $ User
+                        { userName = ""
+                        , userEmail = ""
+                        , userPlugin = credsPlugin creds
+                        , userIdent = credsIdent creds
+                        , userPlan = planId
+                        , userStripeId = Nothing
+                        }
+
+                tryAuthentication creds `shouldReturn` Right userId
+
+                Just user <- runDB $ DB.get userId
+                userName user `shouldBe` "foo"
+                userEmail user `shouldBe` "bar@gmail.com"
+
+            it "errors if name is missing" $ do
+                let creds = Creds
+                        { credsPlugin = "github"
+                        , credsIdent = "1"
+                        , credsExtra = [("email", "foo@gmail.com")]
+                        }
+
+                tryAuthentication creds `shouldReturn` Left "missing key name"
+
+            it "errors if email is missing" $ do
+                let creds = Creds
+                        { credsPlugin = "github"
+                        , credsIdent = "1"
+                        , credsExtra = [("name", "foo")]
+                        }
+
+                tryAuthentication creds `shouldReturn` Left "missing key email"
+
+-- Helper required to disambiguate AuthId m ~ UserId type
```

Comment:
Yes, so the helper is required to fix `m` as `App` because `AuthId m` is too generic to be used as a `UserId` (which is `AuthId App`). It has to be generic for... reasons:

- Fixing it to `UserId` doesn't work b/c `AuthId App ~ UserId` happens in Foundation
- Fixing it to `AuthId App` doesn't work b/c `App` is defined in Foundation
- I can't import Foundation in Model.User b/c that's a cycle

So... a helper or explicit annotation is needed, but I can at least clarify the comment